ydb/library/actors/core/event_pb.cpp

```diff
@@ -221,4 +224,212 @@ namespace NActors {
+            }
+            if (size) {
+                chunker->BackUp(std::exchange(size, 0));
+            }
+            for (const TRope& rope : payload) {
+                if (!chunker->WriteRope(&rope)) {
+                    return false;
+                }
+            }
+#else
+            char marker = PayloadMarker;
+            append(&marker, 1);
+            if (!appendNumber(payload.size())) {
+                return false;
+            }
+            for (const TRope& rope : payload) {
+                if (!appendNumber(rope.GetSize())) {
+                    return false;
+                }
+                if (rope) {
+                    if (size) {
+                        chunker->BackUp(std::exchange(size, 0));
+                    }
+                    if (!chunker->WriteRope(&rope)) {
+                        return false;
+                    }
+                }
+            }
+            if (size) {
+                chunker->BackUp(size);
+            }
+#endif
+        }
+
+        return true;
+    }
+
+
+    void ParseExtendedFormatPayload(TRope::TConstIterator &iter, size_t &size, TVector<TRope> &payload, size_t &totalPayloadSize)
+    {
+        // check marker
+        if (!iter.Valid() || (*iter.ContiguousData() != PayloadMarker && *iter.ContiguousData() != ExtendedPayloadMarker)) {
+            Y_ABORT("invalid event");
+        }
+
+        const bool dataIsSeparate = *iter.ContiguousData() == ExtendedPayloadMarker; // ropes go after sizes
+
+        auto fetchRope = [&](size_t len) {
+            TRope::TConstIterator begin = iter;
+            iter += len;
+            size -= len;
+            payload.emplace_back(begin, iter);
+            totalPayloadSize += len;
+        };
+
+        // skip marker
+        iter += 1;
+        --size;
+        // parse number of payload ropes
+        size_t numRopes = DeserializeNumber(iter, size);
+        if (numRopes == Max<size_t>()) {
+            Y_ABORT("invalid event");
+        }
+        TStackVec<size_t, 16> ropeLens;
+        if (dataIsSeparate) {
+            ropeLens.reserve(numRopes);
+
+        }
+        while (numRopes--) {
+            // parse length of the rope
+            const size_t len = DeserializeNumber(iter, size);
+            if (len == Max<size_t>() || size < len) {
+                Y_ABORT("invalid event len# %zu size# %" PRIu64, len, size);
+            }
+            // extract the rope
+            if (dataIsSeparate) {
+                ropeLens.push_back(len);
+            } else {
+                fetchRope(len);
+            }
+        }
+        for (size_t len : ropeLens) {
+            fetchRope(len);
+        }
+    }
+
+    ui32 CalculateSerializedSizeImpl(const TVector<TRope> &payload, ssize_t recordSize) {
+        ssize_t result = recordSize;
+        if (result >= 0 && payload) {
+            ++result; // marker
+            char buf[MaxNumberBytes];
+            result += SerializeNumber(payload.size(), buf);
```

Comment:
We can compute count of needed bytes faster:
https://github.com/protocolbuffers/protobuf/commit/7315f6d398d2d18443b26c513cbdcdbffaeebaa3

Reply:
```cpp
#if (defined(__x86__) || defined(__x86_64__) || defined(_M_IX86) || \
     defined(_M_X64)) &&                                            \
    !(defined(__LZCNT__) || defined(__AVX2__))
// X86 CPUs lacking the lzcnt instruction are faster with the bsr-based
// implementation. MSVC does not define __LZCNT__, the nearest option that
// it interprets as lzcnt availability is __AVX2__.
#define EVENT_PB_PREFER_BSR 1
#else
#define EVENT_PB_PREFER_BSR 0
#endif

template<typename T>
inline size_t VarintSize(T value) {
#if EVENT_PB_PREFER_BSR
  // Explicit OR 0x1 to avoid calling std::countl_zero(0), which
  // requires a branch to check for on platforms without a clz instruction.
  uint32_t log2value = (std::numeric_limits<T>::digits - 1) -
                       std::countl_zero(value | 0x1);
  return static_cast<size_t>((log2value * 9 + (64 + 9)) / 64);
#else
  uint32_t clz = std::countl_zero(value);
  return static_cast<size_t>(
      ((std::numeric_limits<T>::digits * 9 + 64) - (clz * 9)) / 64);
#endif
}
```